### ISSUE TITLE
storing in a variable results in needless copying

### ISSUE THREAD

**Issue by tspiteri** [rust-lang/rust#48533]
With this *lib.rs*
```rust
pub struct S([i64; 2]);
pub fn foo(create: fn() -> S) -> S {
    create()
}
pub fn bar(create: fn() -> S) -> S {
    let s = create();
    s
}
```
`rustc lib.rs --crate-type lib -C opt-level=3 --emit=asm` produces this *lib.s* (cleaned up)
```asm
foo:
	pushq	%rbx
	movq	%rdi, %rbx
	callq	*%rsi
	movq	%rbx, %rax
	popq	%rbx
	retq
bar:
	pushq	%rbx
	subq	$16, %rsp
	movq	%rdi, %rbx
	movq	%rsp, %rdi
	callq	*%rsi
	movups	(%rsp), %xmm0
	movups	%xmm0, (%rbx)
	movq	%rbx, %rax
	addq	$16, %rsp
	popq	%rbx
	retq
```
The assembly for `bar` should not be longer than the assembly for `foo`.

This is similar to #42870. I don't know if it's a duplicate or just similar.

**Comment by dotdash** [comment]
http://lists.llvm.org/pipermail/llvm-dev/2018-March/121577.html

**Comment by steveklabnik** [comment]
Triage: no change.

**Comment by slightlyoutofphase** [comment]
Currently (and they might have always) those _do_ generate identical assembly when you actually call them in a program. [Playground link to demonstrate.](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=447e307bc4576a82a750ab4b3bbfbce8)

**Comment by tspiteri** [comment]
This is fixed in nightly, possibly due to #72205.
